test/loopback-dl.test.js

```diff
@@ -1421,6 +1421,44 @@ describe('DataAccessObject', function() {
     assert.deepEqual(where, {or: [{age: 10}, {vip: true}]});
   });
 
+  it('should be able to coerce where clauses with array-like objects', function() {
+    where = model._coerce({scores: {0: '10', 1: '20'}});
+    assert.deepEqual(where, {scores: [10, 20]});
+
+    where = model._coerce({and: {0: {age: '10'}, 1: {vip: 'true'}}});
+    assert.deepEqual(where, {and: [{age: 10}, {vip: true}]});
+
+    where = model._coerce({or: {0: {age: '10'}, 1: {vip: 'true'}}});
+    assert.deepEqual(where, {or: [{age: 10}, {vip: true}]});
+
+    where = model._coerce({id: {inq: {0: 'aaa', 1: 'bbb'}}});
+    assert.deepEqual(where, {id: {inq: ['aaa', 'bbb']}});
+
+    where = model._coerce({id: {nin: {0: 'aaa', 1: 'bbb'}}});
+    assert.deepEqual(where, {id: {nin: ['aaa', 'bbb']}});
+
+    where = model._coerce({scores: {between: {0: '0', 1: '42'}}});
+    assert.deepEqual(where, {scores: {between: [0, 42]}});
+  });
+
+  it('should throw on malformed array-like objects for array-only operators', function() {
+    var invalid = [
+      {scores: {inq: {0: '10', 1: '20', 4: '30'}}},
+      {scores: {inq: {0: '10', 1: '20', bogus: 'true'}}},
+      {scores: {between: {0: '10', 1: '20', 2: '30'}}},
+    ];
+
+    invalid.forEach(function(where) {
```

Comment:
Ditto, please split this into multiple tests. You can write a parameterised test case, e.g.

```js
context('malformed array-like objects for array-only operators', function() {
  var INVALID_WHERES = [
    {scores: {inq: {0: '10', 1: '20', 4: '30'}}},
    {scores: {inq: {0: '10', 1: '20', bogus: 'true'}}},
    {scores: {between: {0: '10', 1: '20', 2: '30'}}},
  ];

  INVALID_WHERE.forEach(function(where) {
    it('should reject ' + JSON.stringify(where), function() {
      expect(function() { model._coerce(where) })
        .to.throw(/*...*/);
    });
  });
});
```